File: src/Value.ts
import {InkObject} from './Object';
import {Path} from './Path';
import {InkList, InkListItem} from './InkList';
import {StoryException} from './StoryException';
import {asOrNull, asOrThrows} from './TypeAssertion';

abstract class AbstractValue extends InkObject{
	public abstract get valueType(): ValueType;
	public abstract get isTruthy(): boolean;
	public abstract get valueObject(): any;

	public abstract Cast(newType: ValueType): Value<any>;

	public static Create(val: any): Value<any> | null{
		// Implicitly convert bools into ints
		if (typeof val === 'boolean'){
			let b = !!val;
			val = (b) ? 1 : 0;
		}

		if (Number.isInteger(Number(val))) {
			return new IntValue(Number(val));
		} else if (!isNaN(val)) {
			return new FloatValue(Number(val));
		} else if (typeof val === 'string') {
			return new StringValue(String(val));
		} else if (val instanceof Path) {
			return new DivertTargetValue(asOrThrows(val, Path));
		} else if (val instanceof InkList) {
			return new ListValue(asOrThrows(val, InkList));
		}

		return null;
	}
	public Copy(): InkObject {
		return asOrThrows(AbstractValue.Create(this), InkObject);
	}
	public BadCastException(targetType: ValueType): StoryException {
		return new StoryException("Can't cast "+this.valueObject+' from ' + this.valueType+' to '+targetType);
	}
}

export abstract class Value<T> extends AbstractValue{
	public value: T;

	constructor(val: T){
		super();
		this.value = val;
	}
	public get valueObject(): any{
		return this.value;
	}
	public toString(): string{
		return this.value.toString();
	}
}

export class IntValue extends Value<number>{
	constructor(val: number){
		super(val || 0);
	}
	public get isTruthy(): boolean{
		return this.value != 0;
	}
	public get valueType(): ValueType {
		return ValueType.Int;
	}

	public Cast(newType: ValueType): Value<any>{
		if (newType == this.valueType) {
			return this;
		}

		if (newType == ValueType.Float) {
			return new FloatValue(this.value);
		}

		if (newType == ValueType.String) {
			return new StringValue('' + this.value);
		}

		throw this.BadCastException(newType);
	}
}

export class FloatValue extends Value<number>{
	constructor(val: number){
		super(val || 0.0);
	}
	public get isTruthy(): boolean{
		return this.value != 0.0;
	}
	public get valueType(){
		return ValueType.Float;
	}

	public Cast(newType: ValueType): Value<any>{
		if (newType == this.valueType) {
			return this;
		}

		if (newType == ValueType.Int) {
			return new IntValue(this.value);
		}

		if (newType == ValueType.String) {
			return new StringValue('' + this.value);
		}

		throw this.BadCastException(newType);
	}
}

export class StringValue extends Value<string>{
	public _isNewline: boolean;
	public _isInlineWhitespace: boolean;

	constructor(val: string){
		super(val || '');

		this._isNewline = (this.value == '\n');
		this._isInlineWhitespace = true;

		this.value.split('').every((c) => {
			if (c != ' ' && c != '\t'){
				this._isInlineWhitespace = false;
				return false;
			}

			return true;
		});
	}
	public get valueType(): ValueType{
		return ValueType.String;
	}
	public get isTruthy(): boolean{
		return this.value.length > 0;
	}
	public get isNewline(): boolean{
		return this._isNewline;
	}
	public get isInlineWhitespace(): boolean{
		return this._isInlineWhitespace;
	}
	public get isNonWhitespace(): boolean{
		return !this.isNewline && !this.isInlineWhitespace;
	}

	public Cast(newType: ValueType): Value<any>{
		if (newType == this.valueType) {
			return this;
		}

		if (newType == ValueType.Int) {

			let parsedInt;
			if (parsedInt = parseInt(this.value)) {
				return new IntValue(parsedInt);
			} else {
				throw this.BadCastException(newType);
			}
		}

		if (newType == ValueType.Float) {
			let parsedFloat;
			if (parsedFloat = parseFloat(this.value)) {
				return new FloatValue(parsedFloat);
			} else {
				throw this.BadCastException(newType);
			}
		}

		throw this.BadCastException(newType);
	}
}

export class DivertTargetValue extends Value<Path>{
	constructor(targetPath: Path){
		super(targetPath);

	}
	public get valueType(): ValueType {
		return ValueType.DivertTarget;
	}
	public get targetPath(): Path{
		return this.value;
	}
	public set targetPath(value: Path){
		this.value = value;
	}
	public get isTruthy(): never{
		throw new Error('Shouldn\'t be checking the truthiness of a divert target');
	}

	public Cast(newType: ValueType): Value<any>{
		if (newType == this.valueType)
			return this;

		throw this.BadCastException(newType);
	}
	public toString(): string{
		return 'DivertTargetValue(' + this.targetPath + ')';
	}
}

export class VariablePointerValue extends Value<string>{
	public _contextIndex: number;

	constructor(variableName: string, contextIndex: number = -1){
		super(variableName);

		this._contextIndex = contextIndex;
	}

	public get contextIndex(): number{
		return this._contextIndex;
	}
	public set contextIndex(value: number) {
		this._contextIndex = value;
	}
	public get variableName(): string{
		return this.value;
	}
	public set variableName(value: string){
		this.value = value;
	}
	public get valueType(): ValueType {
		return ValueType.VariablePointer;
	}

	public get isTruthy(): never{
		throw new Error("Shouldn't be checking the truthiness of a variable pointer");
	}

	public Cast(newType: ValueType): Value<any>{
		if (newType == this.valueType)
			return this;

		throw this.BadCastException(newType);
	}
	public toString(): string{
		return 'VariablePointerValue(' + this.variableName + ')';
	}
	public Copy(): InkObject{
		return new VariablePointerValue(this.variableName, this.contextIndex);
	}
}

export class ListValue extends Value<InkList>{
	public get isTruthy(): boolean{
		let isTruthy = false;
		this.value.forEach((value, key, map) => {
			let listItemIntValue = value;
			if (listItemIntValue != 0)
				isTruthy = true;
		});
		return isTruthy;
	}
	public get valueType(): ValueType {
		return ValueType.List;
	}
	public Cast(newType: ValueType): Value<any>{
		 if (newType == ValueType.Int) {
			let max = this.value.maxItem;
			if( max.Key.isNull )
				return new IntValue(0);
			else
				return new IntValue(max.Value);
		}

		else if (newType == ValueType.Float) {
			let max = this.value.maxItem;
			if (max.Key.isNull)
				return new FloatValue(0.0);
			else
				return new FloatValue(max.Value);
		}

		else if (newType == ValueType.String) {
			let max = this.value.maxItem;
			if (max.Key.isNull)
				return new StringValue('');
			else {
				return new StringValue(max.Key.toString());
			}
		}

		 if (newType == this.valueType) return this;

		 throw this.BadCastException(newType);
	}
	// tslint:disable:unified-signatures
	constructor();
	constructor(list: InkList);
	constructor(listOrSingleItem: InkListItem, singleValue: number)
	constructor(listOrSingleItem?: InkListItem | InkList, singleValue?: number){
		if (listOrSingleItem instanceof InkList){
			super(new InkList(listOrSingleItem));
		}
		else if (listOrSingleItem !== undefined && singleValue !== undefined){
			super(new InkList({
				Key: listOrSingleItem,
				Value: singleValue,
			}));
		}
		else{
			super(new InkList());
		}
	}
	public static RetainListOriginsForAssignment(oldValue: InkObject, newValue: InkObject){
		let oldList = asOrNull(oldValue, ListValue);
		let newList = asOrNull(newValue, ListValue);

		// When assigning the empty list, try to retain any initial origin names
		if (oldList && newList && newList.value.Count == 0)
			newList.value.SetInitialOriginNames(oldList.value.originNames);
	}
}

export enum ValueType {
	Int = 0,
	Float = 1,
	List = 2,
	String = 3,
	DivertTarget = 4,
	VariablePointer = 5,
}

File: ink_dart/lib/src/engine/value.dart
import 'ink_object.dart';
import 'path.dart';
import 'ink_list.dart';

/// Enumeration of all Ink value types.
enum ValueType {
  bool_(-1),
  int_(0),
  float_(1),
  list(2),
  string(3),
  divertTarget(4),
  variablePointer(5);

  final int id;
  const ValueType(this.id);
}

/// Exception thrown when a value cannot be cast to the target type.
class BadCastException implements Exception {
  final String message;
  BadCastException(this.message);

  @override
  String toString() => 'BadCastException: $message';
}

/// Abstract base class for all Ink values.
abstract class AbstractValue extends InkObject {
  /// The type of this value.
  ValueType get valueType;

  /// Whether this value is considered "true" in boolean context.
  bool get isTruthy;

  /// The underlying value as a dynamic object.
  Object? get valueObject;

  /// Cast this value to a different type.
  AbstractValue cast(ValueType newType);

  /// Create a value from a Dart object.
  ///
  /// Automatically determines the appropriate Value subclass.
  static AbstractValue? create(Object? val, [ValueType? preferredNumberType]) {
    // Handle preferred number type for explicit typing
    if (preferredNumberType != null) {
      if (preferredNumberType == ValueType.int_ && val is num) {
        if (val == val.toInt()) {
          return IntValue(val.toInt());
        }
      } else if (preferredNumberType == ValueType.float_ && val is num) {
        return FloatValue(val.toDouble());
      }
    }

    if (val is bool) {
      return BoolValue(val);
    }

    if (val is String) {
      return StringValue(val);
    }

    if (val is int) {
      return IntValue(val);
    }

    if (val is double) {
      return FloatValue(val);
    }

    if (val is Path) {
      return DivertTargetValue(val);
    }

    if (val is InkList) {
      return ListValue.fromList(val);
    }

    return null;
  }

  @override
  InkObject copy() {
    final created = AbstractValue.create(valueObject);
    if (created == null) {
      throw StateError('Could not copy value: $valueObject');
    }
    return created;
  }

  /// Create a BadCastException for invalid casts.
  BadCastException badCastException(ValueType targetType) {
    return BadCastException(
      "Can't cast $valueObject from $valueType to $targetType",
    );
  }
}

/// Generic base class for typed values.
abstract class Value<T> extends AbstractValue {
  T? value;

  Value(this.value);

  @override
  Object? get valueObject => value;

  @override
  String toString() {
    if (value == null) {
      throw StateError('Value is null');
    }
    return value.toString();
  }
}

/// Boolean value.
class BoolValue extends Value<bool> {
  BoolValue(bool val) : super(val);

  @override
  bool get isTruthy => value ?? false;

  @override
  ValueType get valueType => ValueType.bool_;

  @override
  AbstractValue cast(ValueType newType) {
    if (value == null) throw StateError('Value is null');

    if (newType == valueType) return this;

    switch (newType) {
      case ValueType.int_:
        return IntValue(value! ? 1 : 0);
      case ValueType.float_:
        return FloatValue(value! ? 1.0 : 0.0);
      case ValueType.string:
        return StringValue(value! ? 'true' : 'false');
      default:
        throw badCastException(newType);
    }
  }

  @override
  String toString() => value == true ? 'true' : 'false';
}

/// Integer value.
class IntValue extends Value<int> {
  IntValue(int val) : super(val);

  @override
  bool get isTruthy => (value ?? 0) != 0;

  @override
  ValueType get valueType => ValueType.int_;

  @override
  AbstractValue cast(ValueType newType) {
    if (value == null) throw StateError('Value is null');

    if (newType == valueType) return this;

    switch (newType) {
      case ValueType.bool_:
        return BoolValue(value != 0);
      case ValueType.float_:
        return FloatValue(value!.toDouble());
      case ValueType.string:
        return StringValue(value.toString());
      default:
        throw badCastException(newType);
    }
  }
}

/// Floating-point value.
class FloatValue extends Value<double> {
  FloatValue(double val) : super(val);

  @override
  bool get isTruthy => (value ?? 0.0) != 0.0;

  @override
  ValueType get valueType => ValueType.float_;

  @override
  AbstractValue cast(ValueType newType) {
    if (value == null) throw StateError('Value is null');

    if (newType == valueType) return this;

    switch (newType) {
      case ValueType.bool_:
        return BoolValue(value != 0.0);
      case ValueType.int_:
        return IntValue(value!.toInt());
      case ValueType.string:
        return StringValue(value.toString());
      default:
        throw badCastException(newType);
    }
  }
}

/// String value.
class StringValue extends Value<String> {
  late final bool _isNewline;
  late final bool _isInlineWhitespace;

  StringValue(String val) : super(val) {
    _isNewline = value == '\n';
    _isInlineWhitespace = _computeIsInlineWhitespace();
  }

  bool _computeIsInlineWhitespace() {
    if (value == null || value!.isEmpty) {
      return true;
    }
    for (final c in value!.split('')) {
      if (c != ' ' && c != '\t') {
        return false;
      }
    }
    return true;
  }

  @override
  ValueType get valueType => ValueType.string;

  @override
  bool get isTruthy => (value?.isNotEmpty ?? false);

  /// Whether this string is just a newline.
  bool get isNewline => _isNewline;

  /// Whether this string is only spaces and tabs.
  bool get isInlineWhitespace => _isInlineWhitespace;

  /// Whether this string has non-whitespace content.
  bool get isNonWhitespace => !isNewline && !isInlineWhitespace;

  @override
  AbstractValue cast(ValueType newType) {
    if (newType == valueType) return this;

    if (newType == ValueType.int_) {
      final parsed = int.tryParse(value ?? '');
      if (parsed != null) {
        return IntValue(parsed);
      }
      throw badCastException(newType);
    }

    if (newType == ValueType.float_) {
      final parsed = double.tryParse(value ?? '');
      if (parsed != null) {
        return FloatValue(parsed);
      }
      throw badCastException(newType);
    }

    throw badCastException(newType);
  }
}

/// Divert target value - a path to a location in the story.
class DivertTargetValue extends Value<Path> {
  DivertTargetValue([Path? targetPath]) : super(targetPath);

  @override
  ValueType get valueType => ValueType.divertTarget;

  /// The target path.
  Path get targetPath {
    if (value == null) throw StateError('Value is null');
    return value!;
  }

  set targetPath(Path path) {
    value = path;
  }

  @override
  bool get isTruthy {
    throw StateError("Shouldn't be checking the truthiness of a divert target");
  }

  @override
  AbstractValue cast(ValueType newType) {
    if (newType == valueType) return this;
    throw badCastException(newType);
  }

  @override
  String toString() => 'DivertTargetValue($targetPath)';
}

/// Variable pointer value - a reference to a variable.
class VariablePointerValue extends Value<String> {
  int _contextIndex;

  VariablePointerValue(String variableName, [int contextIndex = -1])
      : _contextIndex = contextIndex,
        super(variableName);

  @override
  ValueType get valueType => ValueType.variablePointer;

  /// The context index for variable lookup.
  int get contextIndex => _contextIndex;
  set contextIndex(int value) => _contextIndex = value;

  /// The variable name.
  String get variableName {
    if (value == null) throw StateError('Value is null');
    return value!;
  }

  set variableName(String name) {
    value = name;
  }

  @override
  bool get isTruthy {
    throw StateError("Shouldn't be checking the truthiness of a variable pointer");
  }

  @override
  AbstractValue cast(ValueType newType) {
    if (newType == valueType) return this;
    throw badCastException(newType);
  }

  @override
  String toString() => 'VariablePointerValue($variableName)';

  @override
  InkObject copy() => VariablePointerValue(variableName, contextIndex);
}

/// List value - a set of named items from list definitions.
class ListValue extends Value<InkList> {
  ListValue() : super(InkList());

  ListValue.fromList(InkList list) : super(InkList.from(list));

  ListValue.fromItem(InkListItem item, int itemValue)
      : super(InkList.fromItem(item, itemValue));

  @override
  bool get isTruthy {
    if (value == null) throw StateError('Value is null');
    return value!.isNotEmpty;
  }

  @override
  ValueType get valueType => ValueType.list;

  @override
  AbstractValue cast(ValueType newType) {
    if (value == null) throw StateError('Value is null');

    if (newType == ValueType.int_) {
      final max = value!.maxItem;
      if (max == null || max.key.isNull) return IntValue(0);
      return IntValue(max.value);
    }

    if (newType == ValueType.float_) {
      final max = value!.maxItem;
      if (max == null || max.key.isNull) return FloatValue(0.0);
      return FloatValue(max.value.toDouble());
    }

    if (newType == ValueType.string) {
      final max = value!.maxItem;
      if (max == null || max.key.isNull) return StringValue('');
      return StringValue(max.key.toString());
    }

    if (newType == valueType) return this;

    throw badCastException(newType);
  }

  /// Retain origin names when assigning empty list.
  static void retainListOriginsForAssignment(
    InkObject? oldValue,
    InkObject newValue,
  ) {
    if (oldValue is! ListValue || newValue is! ListValue) return;

    final oldList = oldValue.value;
    final newList = newValue.value;

    if (oldList == null || newList == null) return;

    // When assigning the empty list, try to retain any initial origin names
    if (newList.isEmpty) {
      newList.setInitialOriginNames(oldList.originNames);
    }
  }
}

